src/instrumentation/do-rpc.ts

```diff
@@ -0,0 +1,249 @@
+		span.setAttribute(SemanticAttributes.FAAS_COLDSTART, cold_start)
+		cold_start = false
+		span.setAttribute('do.id', id.toString())
+		if (id.name) span.setAttribute('do.name', id.name)
+
+		try {
+			await alarmFn()
+			span.end()
+		} catch (error) {
+			span.recordException(error as Exception)
+			span.setStatus({ code: SpanStatusCode.ERROR })
+			span.end()
+			throw error
+		}
+	})
+	return promise
+}
+
+function instrumentDORPCFn(anyFn: AnyFn, fnName: string, initialiser: Initialiser, env: Env, id: DurableObjectId): AnyFn {
+	const fetchHandler: ProxyHandler<AnyFn> = {
+		async apply(target, thisArg, argArray: any) {
+			const request = argArray[0]?.request;
+			const config = initialiser(env, request)
+			const context = setConfig(config)
+			try {
+				const bound = target.bind(unwrap(thisArg))
+				let response = await api_context.with(context, executeDORPCFn, undefined, bound, fnName, argArray, id)
+				return response
+			} catch (error) {
+				throw error
+			} finally {	
+				exportSpans()
+			}
+		},
+	}
+	return wrap(anyFn, fetchHandler)
+}
+
+function instrumentAlarmFn(alarmFn: AlarmFn, initialiser: Initialiser, env: Env, id: DurableObjectId) {
+	if (!alarmFn) return undefined
+
+	const alarmHandler: ProxyHandler<NonNullable<AlarmFn>> = {
+		async apply(target, thisArg) {
+			const config = initialiser(env, 'do-alarm')
+			const context = setConfig(config)
+			try {
+				const bound = target.bind(unwrap(thisArg))
+				return await api_context.with(context, executeDOAlarm, undefined, bound, id)
+			} catch (error) {
+				throw error
+			} finally {
+				exportSpans()
+			}
+		},
+	}
+	return wrap(alarmFn, alarmHandler)
+}
+
+function instrumentDurableObjectStub(doObj: DurableObject, initialiser: Initialiser, env: Env, state: DurableObjectState, rpcFunctions: string[]) {
+	const objHandler: ProxyHandler<DurableObject> = {
+		get(target, prop) {
+			if (prop === 'alarm') {
+				const alarmFn = Reflect.get(target, prop)
+				return instrumentAlarmFn(alarmFn, initialiser, env, state.id)
+			} else {
+				const result = Reflect.get(target, prop)
+				if (typeof result === 'function' && rpcFunctions.includes(String(prop))) {
+					return instrumentDORPCFn(result, String(prop), initialiser, env, state.id)
+				}
+				return result
+			}
+		},
+	}
+	return wrap(doObj, objHandler)
+}
+
+export function instrumentDOClassRPC(doClass: DOClass, initialiser: Initialiser, rpcFunctions: string[]): DOClass {
+	const classHandler: ProxyHandler<DOClass> = {
+		construct(target, [orig_state, orig_env]: ConstructorParameters<DOClass>) {
+			const trigger: DOConstructorTrigger = {
+				id: orig_state.id.toString(),
+				name: orig_state.id.name,
+			}
+			const constructorConfig = initialiser(orig_env, trigger)
+			const context = setConfig(constructorConfig)
+			const env = instrumentEnv(orig_env)
+			const createDO = () => {
+				orig_state.storage.sql = instrumentSql(orig_state.storage.sql)
```

Comment:
DurableObject constructor asserts that the state is a DurableObjectState, so passing a ProxyHandler doesn't work. Thus this hack.